### PR TITLE
[#12230][fix] Add bounds checking in autotuner _find_nearest_profile for SM121

### LINKED ISSUE #12230: [Bug]: Qwen3-Next-80B FP8/NVFP4 — Autotuner IndexError and Triton Illegal Instruction on DGX Spark (aarch64 Blackwell, TRT-LLM 1.3.0rc7)
### System Info

- CPU architecture: aarch64 (ARM Grace)
- CPU/Host memory size: 128 GB unified (shared with GPU)
- GPU name: NVIDIA GB10 Blackwell (DGX Spark)
- GPU memory size: 128 GB unified
- TensorRT-LLM version: 1.3.0rc7
- Container: nvcr.io/nvidia/tensorrt-llm/release:1.3.0rc7
- PyTorch version: 2.10.0a0+b4e4ee8 (NV25.12)
- CUDA version: 13.1 (driver 580.126.09, forward compat)
- Models tested:
    - Qwen/Qwen3-Next-80B-A3B-Thinking-FP8
    - nvidia/Qwen3-Next-80B-A3B-Thinking-NVFP4

### Who can help?

@Tracin (FP8 quantization / autotuner)
@kaiyux (performance / kernel)

### Information

- [ ] The official example scripts
- [x] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [x] My own task or dataset (give details below)

### Reproduction

My own task: serving Qwen3-Next-80B (FP8 and NVFP4) via trtllm-serve
on NVIDIA DGX Spark GB10 (aarch64, 128 GB unified memory).

### Bug 1: FP8 — Autotuner IndexError (crashes at warmup)

**Command:**
trtllm-serve "/app/models/Qwen3-Next-80B-A3B-Thinking-FP8" \
  --host 0.0.0.0 --port 8355 \
  --backend pytorch \
  --extra_llm_api_options /app/config/extra-config.yml

**Config (extra-config.yml):**
print_iter_log: false
max_batch_size: 2
max_num_tokens: 49152
kv_cache_config:
  free_gpu_memory_fraction: 0.95
  enable_block_reuse: false     # required: Mamba hybrid arch
cuda_graph_config: null         # required: ARM + Mamba crashes graph capture
attn_backend: TRTLLM
moe_config:
  backend: CUTLASS
disable_overlap_scheduler: true
enable_chunked_prefill: false

**Error:**
[TRT-LLM] [I] Use 74.85 GB for model weights.
Loading weights concurrently: 100%|██████████| 1412/1412 [04:35<00:00]
Model init total -- 315.92s
[TRT-LLM] [I] Running autotuner warmup...
[TRT-LLM] [I] [Autotuner] Autotuning process starts ...
[TRT-LLM] [E] Failed to initialize executor on rank 0: list assignment index out of range

Traceback (most recent call last):
  File "tensorrt_llm/_torch/autotuner.py", line 1406, in _find_nearest_profile
    base_profile[spec.input_idx][spec.dim_idx] = -1
    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^
IndexError: list assignment index out of range

Call path:
  trtllm-serve
    → _run_autotuner_warmup
      → modeling_qwen3_next.py forward
        → in_proj_qkvz (linear_attn)
          → fp8_block_scaling_gemm
            → tuner.choose_one
              → profiling_cache.search_cache
                → _find_nearest_profile   ← IndexError here

---

### Bug 2: NVFP4 — Triton Illegal Instruction on ARM (crashes at first forward)

**Command:** same as above, model path changed to NVFP4 variant.

**Error:**
RuntimeError: Triton Error [CUDA]: an illegal instruction was encountered

  File "tensorrt_llm/_torch/models/modeling_qwen3_next.py", line 332:
      fused_qkvzba_split_reshape_cat_kernel[grid](

The custom Triton kernel fused_qkvzba_split_reshape_cat_kernel in
modeling_qwen3_next.py produces an illegal instruction on
ARM aarch64 + GB10 (SM121). The kernel appears compiled for x86/SM10x
and is not compatible with the aarch64 execution environment.

### Expected behavior

Both variants (FP8 and NVFP4) of Qwen3-Next-80B should serve successfully
via trtllm-serve on DGX Spark GB10 (Blackwell SM121, aarch64).

- FP8: the autotuner should complete warmup or fall back gracefully to a
  default CUTLASS tactic instead of raising an unhandled IndexError.
- NVFP4: fused_qkvzba_split_reshape_cat_kernel should execute correctly
  on aarch64 + SM121, or a CPU/non-Triton fallback path should exist.

### actual behavior

- FP8: model loads successfully (74.85 GB, 1412 shards, ~316 s),
  but the server crashes during autotuner warmup with:
  IndexError: list assignment index out of range
  No inference is possible; the server never starts.

- NVFP4: crashes immediately on the first forward pass with:
  RuntimeError: Triton Error [CUDA]: an illegal instruction was encountered
  in fused_qkvzba_split_reshape_cat_kernel.

Both bugs are specific to TRT-LLM. The FP8 model serves correctly
via vLLM (nvcr.io/nvidia/vllm:26.01-py3) at ~44 t/s on the same hardware,
confirming the weights and hardware are functional.

### additional notes

1. Relation to issue #10679:
   Bug 1 (FP8 autotuner IndexError) is the same _find_nearest_profile crash
   reported in #10679 (MoE + FP8 on H100/x86). That issue is still
   "waiting for feedback" with no fix in 1.3.0rc7.
   This report adds the aarch64 / DGX Spark / GB10 context.

2. pytorch_backend_config is rejected in 1.3.0rc7:
   Both the nested YAML block (pytorch_backend_config:) and top-level
   fields (autotuner_enabled, use_cuda_graph) are rejected by trtllm-serve:
     ValueError: LLM got invalid argument: pytorch_backend_config
   There is currently no documented way to disable the autotuner via config.
   TLLM_DISABLE_AUTOTUNER=1 env var also has no visible effect.

3. Additional config constraints required on DGX Spark:
   - enable_block_reuse: false  → Qwen3-Next Mamba hybrid arch requires this
   - cuda_graph_config: null    → ARM + Mamba combination causes illegal
                                  instruction during CUDA graph capture

4. Questions for the team:
   Q1. Is there a working env var or YAML flag to fully disable the autotuner
       in 1.3.0rc7?
   Q2. Is fused_qkvzba_split_reshape_cat_kernel compiled for aarch64 + SM121?
       If not, is there a non-Triton fallback path?
   Q3. Is Qwen3-Next-80B (FP8 or NVFP4) officially supported on DGX Spark
       GB10 with TRT-LLM? If so, what is the recommended serving config?

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and checked the [documentation](https://nvidia.github.io/TensorRT-LLM/) and [examples](https://github.com/NVIDIA/TensorRT-LLM/tree/main/examples) for answers to frequently asked questions.

### PR DESCRIPTION
**Pull request by mihai-chiorean**
## Summary

Fixes https://github.com/NVIDIA/TensorRT-LLM/issues/12230 (autotuner crash)

On DGX Spark (SM121), `_find_nearest_profile()` crashes with `IndexError: list assignment index out of range` during warmup, preventing the server from starting. This happens because `DynamicTensorSpec.input_idx` or `dim_idx` exceeds the `base_profile` dimensions when ops produce fewer or differently-shaped tensors than the specs expect.

Adds bounds checks before indexing into `base_profile` for both `DynamicTensorSpec` and `ConstraintSpec` loops. Out-of-bounds specs are skipped with a debug log. All in-bounds behavior is unchanged.

## Test Coverage

Tested with 8 cases inside the Docker container:
- Normal operation (in-bounds specs) — unchanged behavior
- `input_idx` out of range for DynamicTensorSpec — skipped cleanly
- `dim_idx` out of range for DynamicTensorSpec — skipped cleanly
- `input_idx`/`dim_idx` out of range for ConstraintSpec — skipped cleanly
- Mixed valid + out-of-bounds specs — valid ones still applied correctly
- `tune_max_num_tokens` with out-of-bounds spec — handled
- Empty shapes with non-empty specs — handled

## Test plan

- [x] 8 unit test cases pass (bounds checking, mixed valid/invalid, edge cases)
- [x] Pre-commit checks pass

Note: Full autotuner integration test requires model serving, deferred to NVIDIA CI.

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- Update [tava architecture diagram](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/tava_architecture_diagram.md) if there is a significant design change in PR.
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

## Summary by CodeRabbit

* **Bug Fixes**
  * Enhanced autotuner robustness with defensive bounds validation to prevent out-of-bounds access errors during profile matching operations.

### CODE REVIEW

File: tensorrt_llm/_torch/autotuner.py [hunk 1]
```diff
@@ -1301,6 +1301,17 @@ def _optimization_profiles(
         for spec in tuning_config.dynamic_tensor_specs:
             assert callable(spec.gen_tuning_buckets) or isinstance(spec.gen_tuning_buckets, (list, tuple)), \
                 "The given dynamic dimension must provide a opt value generation function or a list of opt values"
+            if spec.input_idx >= len(base_profile.shapes):
+                logger.debug(
```
[the remark this answers is not in the document]
Comment (mihai-chiorean) [reply tensorrt_llm/_torch/autotuner.py]: fair point. I made them warnings.

**Comment by mihai-chiorean** [comment]
@hyukn @pamelap-nvidia I fixed the suggestions and a couple of more small things. hopefully we're good this time 🤞

**Review by pamelap-nvidia — APPROVED** [review]

**Review by hyukn — COMMENTED** [review]
Thanks a lot for the contribution! Just left some comments for discussion.

**Comment by mihai-chiorean** [comment]
> Thanks a lot for the contribution! Just left some comments for discussion.

Any other thoughts on this?

**Comment by mihai-chiorean** [comment]
> /bot run --disable-fail-fast

Thanks for triggering. I guess this needs some updates. I can't see the test failures but I'll dig through and see what I'm missing

**Comment by mihai-chiorean** [comment]
@karljang @hyukn any chance I can get another ci run and review? this should be good to go now.

**Comment by karljang** [comment]
Thanks for the update~ 👍

**Comment by mihai-chiorean** [comment]
> [PR_Github #47809](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/47809/) [ run ] completed with state `SUCCESS`. Commit: `b389fa4` [/LLM/main/L0_MergeRequest_PR pipeline #37699](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/37699/) completed with status: 'FAILURE'
> 
> [CI Report](http://tensorrt-llm.tensorrt-llm-ci-report.sc2-paas.nvidia.com/?job=%2FLLM%2Fmain%2FL0_MergeRequest_PR&build=37699)
> 
> ⚠️ **Action Required:**
> 
> * Please check the failed tests and fix your PR
> * If you cannot view the failures, ask the CI triggerer to share details
> * Once fixed, request an NVIDIA team member to trigger CI again
> 
> [CI Agent Failure Analysis](https://pbss.s8k.io/v1/AUTH_svc_tensorrt/sw-tensorrt-ci-analysis/LLM/main/L0_MergeRequest_PR/37699/failure_analysis.html)
> 
> [Link to invocation](https://github.com/NVIDIA/TensorRT-LLM/pull/12310#issuecomment-4426377228)

I wish I could see why this fails.

**Comment by karljang** [comment]
Sure, my ClaudeCode found some issues. I'll let it post comments.

**Comment by karljang** [comment]
Thanks for the fix! Heads up — CI run #37699 has 20 failures that need a look:

## 1. New bounds-check tests in this PR (16 failures, both B200 and H100)

All `TestFindNearestProfileBounds` / `TestOptimizationProfilesBounds` cases fail with the same pattern:

```text
tests/unittest/_torch/misc/test_autotuner.py:1028
    assert any("Skipping DynamicTensorSpec with input_idx=5" in m ...)
E   assert False
```

The generator finds zero matches in captured logs, so either:

- (a) the bounds-check path in `autotuner.py` isn't logging when it skips,
- (b) the message text doesn't match what the test searches for, or
- (c) `caplog` isn't capturing at the right level / logger.

Since every parametrization fails identically, one root cause should fix all 16.

Quick local repro:

```bash
pytest tests/unittest/_torch/misc/test_autotuner.py::TestFindNearestProfileBounds \
  -v -s --log-cli-level=DEBUG
```

## 2. Two unrelated MoE failures on B200 (probably flaky)

```text
test_moe_backend.py::test_moe_backend[...backend=CUTLASS...routing=Renormalize]
  → Test terminated unexpectedly
```

Empty stdout/stderr suggests a worker crash, and this PR doesn't touch MoE — likely a flake. Should clear on the next `/bot run` once the autotuner tests are fixed.

**Comment by mihai-chiorean** [comment]
> [PR_Github #50295](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/50295/) [ run ] completed with state `FAILURE`. Commit: `c9e005a` [/LLM/main/L0_MergeRequest_PR pipeline #39825](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/39825/) completed with status: 'FAILURE'
> 
> [CI Report](http://tensorrt-llm.tensorrt-llm-ci-report.sc2-paas.nvidia.com/?job=%2FLLM%2Fmain%2FL0_MergeRequest_PR&build=39825)
> 
> ⚠️ **Action Required:**
> 
> * Please check the failed tests and fix your PR
> * If you cannot view the failures, ask the CI triggerer to share details
> * Once fixed, request an NVIDIA team member to trigger CI again
> 
> [CI Agent Failure Analysis](https://pbss.s8k.io/v1/AUTH_svc_tensorrt/sw-tensorrt-ci-analysis/LLM/main/L0_MergeRequest_PR/39825/failure_analysis.html)
> 
> [Link to invocation](https://github.com/NVIDIA/TensorRT-LLM/pull/12310#issuecomment-4540925646)

I'd really love to fix this but I don't know what it is, can't see it. locally on sm121 tests pass.

**Comment by mihai-chiorean** [comment]
> [PR_Github #51158](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/51158/) [ run ] completed with state `SUCCESS`. Commit: `eb548eb` [/LLM/main/L0_MergeRequest_PR pipeline #40592](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/40592/) completed with status: 'SUCCESS'
> 
> [CI Report](http://tensorrt-llm.tensorrt-llm-ci-report.sc2-paas.nvidia.com/?job=%2FLLM%2Fmain%2FL0_MergeRequest_PR&build=40592)
> 
> [Link to invocation](https://github.com/NVIDIA/TensorRT-LLM/pull/12310#issuecomment-4581612138)

@hyukn @yizhang-nv this now has a clean CI pass. any objections to merging it?

File: tensorrt_llm/_torch/autotuner.py [hunk 2]
```diff
@@ -1395,6 +1419,22 @@ def _find_nearest_profile(
         base_profile = list(list(shape) for shape in shapes)
 
         for spec in dynamic_tensor_specs:
+            # Bounds check: skip specs that reference inputs or dimensions not present in the
+            # current shapes tuple. This can happen on hardware (e.g. SM121 / DGX Spark) where
+            # ops produce fewer or differently-shaped tensors than the specs were authored for.
+            if spec.input_idx < 0 or spec.input_idx >= len(base_profile):
```
[the remark this answers is not in the document]
Comment (hyukn) [reply tensorrt_llm/_torch/autotuner.py]: Hi @mihai-chiorean. Sorry for the late response and thanks for the reply.
The bound limit checking is very reasonable and could help us to avoid invalid tuning issues. According to your changes, I have several concerns:
* Can we do this only during the tuning stage, and let it fallback due to cache miss during find_nearest_profile to avoid runitme overhead?
* Can we make the checking in one place for a util method to keep the main block more concrete?
* Maybe we can compress the unit test to just a single.

[the remark this answers is not in the document]
Comment (mihai-chiorean) [reply tensorrt_llm/_torch/autotuner.py]: hey [hyukn](https://github.com/hyukn), thanks for taking a look!

> Can we make the checking in one place for a util method to keep the main block more concrete?
absolutely. I was thinking about that too. done. let me know if this works.

> Maybe we can compress the unit test to just a single.
done. not quite a single, but compressed. hope that works. 

> Can we do this only during the tuning stage, and let it fallback due to cache miss during find_nearest_profile to avoid runitme overhead?

doesn't the bounds check actually enable the cache miss fallback?  without it, an OOB spec triggers an IndexError during key construction — that's the original crash. and the exception propagates up before search_cache can return the miss signal. With the bounds check, the function returns a valid (but un-tuned) key → cache miss → fallback as designed.

on overhead: `_find_nearest_profile` is decorated `@lru_cache(maxsize=None)`, so the bounds check runs once per unique `(shapes, specs)` tuple and zero on the hot path. am I missing something on the cost side?

if the underlying concern is the duplication of the check between `_optimization_profiles` (tune time) and `_find_nearest_profile` (key construction), I can refactor so `choose_one` filters the spec list once at the entry point and both downstream functions then trust their inputs. would that address it? happy to do that as a follow-up commit. but any other callers of `_find_nearest_profile` would have the same bounds check issue.

**Comment by karljang** [comment]
Could you please address the pre-commit failures, @mihai-chiorean ?
Thank you~

**Comment by mihai-chiorean** [comment]
> Could you please address the pre-commit failures, @mihai-chiorean ?
> 
> Thank you~
> 
> 

Oops. Done!

File: tensorrt_llm/_torch/autotuner.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (hyukn) [reply tensorrt_llm/_torch/autotuner.py]: I think this should be the correct way to drop the opportunity for those invalid specs defined in the tuning config. If those shapes can be narrowed to the ones covered by the warmup, then the tuning results can still be reused, otherwise, the cache will always miss for the whole shape list because we only record those shapes with invalid specs showing during warmup.

Actually, I still believe that the shape mismatch between the tuning config and real input tensor is a "wrong usage". But this PR truly looses the constraints so that those op can run with at least the fallback tactic.

Thanks for the explanations. The PR LGTM. Just some minor changes may be required.

File: tensorrt_llm/_torch/autotuner.py [hunk 3]
```diff
@@ -710,6 +710,36 @@ def _deserialize_cache_data(
         return cache
 
 
+def _spec_in_bounds(spec, shapes_list, spec_kind: str) -> bool:
```
Comment (hyukn) [inline tensorrt_llm/_torch/autotuner.py]: Let us directly extract the class type of the spec instead of a standalone str.

File: tests/unittest/_torch/misc/test_autotuner.py [hunk 4]
```diff
@@ -986,3 +986,100 @@ def forward(self,
               f"{event_ms:>16.4f} {gt_ms:>17.4f} "
               f"{abs_diff:>14.4f} {rel_diff * 100:>13.2f}")
     print("-" * 102)
+
+
+# ---------------------------------------------------------------------------
+# Bounds-checking tests (from PR #12310)
```
Comment (hyukn) [inline tests/unittest/_torch/misc/test_autotuner.py]: No need to put PR number here. And I suggest compressing the number of unittests to 1 to cover spec bound func happens in both opt_profile and find_nearest.

**Review by hyukn — COMMENTED** [review]
LGTM, only some minor changes required. Thanks for the effort!

**Comment by mihai-chiorean** [comment]
> LGTM, only some minor changes required. Thanks for the effort!

done

**Review by hyukn — APPROVED** [review]

**Comment by mihai-chiorean** [comment]
thanks @hyukn @pamelap-nvidia @karljang !! really appreciate your patience and reviews! looks like this is now ready.

**Comment by karljang** [comment]
Yeah, finally! Thank you so much for your contribution!
